README


This archive contains the distribution Lingua-Identifier,
version 0.01_8:

  A NN based approach for language identification

This software is copyright (c) 2014 by Alberto Simões.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Lingua-Identifier

0.01_8  2014-08-28
        - SWIG code doesn't deal well it tainted variables.
          Workaround is not taint anything.

0.01_7  2014-08-27
        - Fixed identify_language script to use new API.
        - Fixed dependencies issues.
        - Still debugging...

0.01_6  2014-08-26
        - Added more debug messages :'(

0.01_5  2014-08-25
        - Added debug messages.
        - Bumped Math::Matrix::MaybeGSL version dependency.

0.01_4  2014-08-24
        - Basic API is now stable.
        - Build based on Dist::Zilla.
        - Still wondering what is going on with perl 5.16.3.

0.01_3  2014-08-23
        - Some CPAN testers did not test because of a warning that is now fixed.
        - Added some more documentation, and a public method.

0.01_2  2014-08-22
        - First working version, without documentation, for cpan-testers runs.



inc/MyBuilder.pm
package MyBuilder;
use base 'Module::Build';
use warnings;
use strict;

use autodie;

use Math::Matrix::MaybeGSL;
use Data::Dumper;

sub ACTION_code {
    my $self = shift;

    die "Can not find share dir..." unless -d "share";

    $self->SUPER::ACTION_code;

    $self->dispatch("prepare_features")
      if ! -f "share/features.lst" || -M "share/features.lst" > -M "data/features.dmp";

    $self->dispatch("prepare_classes")
      if ! -f "share/classes.lst"  || -M "share/classes.lst"  > -M "data/classes.dmp";

    $self->dispatch("prepare_thetas")
      if ! -f "share/theta-01.dat" || -M "share/theta-01.dat" > -M "data/thetas.dat";

}

sub ACTION_prepare_thetas {
    my $self = shift;

    print STDERR "Munging some data...";

    open my $fh, "<", "data/thetas.dat";

    # skip comments
    my $line = <$fh>;
    $line = <$fh> while $line =~ /^\s*#/;

    # first line is the number of layers
    chomp(my $nr_layers = $line);

    # then we have the arch
    my @arch;
    for (1..$nr_layers) {
        chomp($line = <$fh>);
        $arch[$_] = $line;
    }

    ## Load and save matrices
    for my $l (1 .. $nr_layers-1) {
        my $m = Matrix->new($arch[$l+1], $arch[$l]+1);
        for my $c (1 .. $arch[$l]+1) {
            for my $r (1 .. $arch[$l+1]) {
                chomp($line = <$fh>);
                $m->assign($r, $c, $line);
            }

            print STDERR "." unless $c % 100;  ## show some progress
        }
        $m->write(sprintf("share/theta-%02d.dat", $l));
    }

    print STDERR "done\n";
}

sub ACTION_prepare_classes {
    my $self = shift;

    open my $fh, "<:utf8", "data/classes.lst";

    my $classes;
    ### We expect them to be ordered, consecutive, starting at 1
    while (<$fh>) {
        chomp;
        my ($id, $class) = split /\s+/;
        push @$classes, $class;
    }

    close $fh;

    open $fh, ">:utf8", "share/classes.dmp";
    print $fh "use utf8;\n";
    print $fh Dumper($classes);
    close $fh;

}

sub ACTION_prepare_features {
    my $self = shift;

    open my $fh, "<:utf8", "data/features.lst";

    my $features;
    while (<$fh>) {
        chomp;
        push @$features, $_;
    }

    close $fh;

    open $fh, ">:utf8", "share/features.dmp";
    print $fh "use utf8;\n";
    print $fh Dumper($features);
    close $fh;
}




1;


lib/Lingua/Identifier.pm
package Lingua::Identifier;
$Lingua::Identifier::VERSION = '0.01_8';
use 5.014004;
use strict;
use warnings FATAL => 'all';

use File::ShareDir 'dist_dir';
use File::Spec::Functions;

use Math::Matrix::MaybeGSL 0.006;

use Lingua::Identifier::ForwardProp;
use Lingua::Identifier::Feature::Trigrams;
use Lingua::Identifier::Feature::Alphabet;

=encoding UTF-8

=head1 NAME

Lingua::Identifier - A NN based approach for language identification

=cut

if (Matrix->isGSL) {
    eval {
        require Math::GSL;
        print STDERR "\n\n Running GSL version ", Math::GSL::gsl_version(), "\n\n";
    }
}

our $sharedir = dist_dir('Lingua-Identifier');

our $features = do( catfile($sharedir, "features.dmp"));
die __PACKAGE__ . "- could not load 'features.dmp'" unless defined $features;

our $classes = do( catfile($sharedir, "classes.dmp"));
die __PACKAGE__ . "- could not load 'classes.dmp'" unless defined $classes;

our $thetas;
_load_thetas($sharedir);


=head1 SYNOPSIS

    use Lingua::Identifier;

    my $identifier = Lingua::Identifier->new();

    # identify language on a file
    my $lang = $identifier->identify_file("text.txt");

    # identify language on a string
    my $lang = $identifier->identify($string);

=head1 DESCRIPTION

This documentation is not ready yet. These releases are just for
CPANtesters testing.

=head2 C<new>

Constructs a new Language Identification object.

    my $identifier = Lingua::Identifier->new();

=cut

sub new {

    return bless { languages => $classes }, __PACKAGE__;
}

=head2 C<languages>

Returns the list of codes for the active languages.

=cut

sub languages {
    my $self = shift;
    return @{$self->{languages}};
}

=head2 C<identify_file>

This method receives a filename and tries to identify its langauge.

In scalar context returns the language id. In list context returns an
associative array, with language codes and respective scores.

    my $lang = $identifier->identify_file("sometext.txt");

=cut

sub identify_file {
    my ($self, $filename) = @_;

    my $string = _load_file($filename);
    $self->identify($string);
}

=head2 C<identify>

This method receives a string and tries to identify its langauge.

In scalar context returns the language id. In list context returns an
associative array, with language codes and respective scores.

    my $lang = $identifier->identify($string);

=cut

sub identify {
    my ($self, $string) = @_;

    my $ngrams = _compute_features($string);

    my $data = Matrix->new(scalar(@$features), 1);

    my $i = 1;
    for my $feature (@$features) {
        if (exists($ngrams->{$feature})) {
            $data->assign($i, 1, $ngrams->{$feature});
        }
        $i++;
    }

    my $ans = Lingua::Identifier::ForwardProp::forward_prop($data, $thetas);

    my ($max, $pos) = $ans->max();

    if (wantarray) {
        my $prob_classes = {};
        my $i = 1;
        for (@$classes) {
            $prob_classes->{$_} = $ans->element($i++, 1);
        }
        return (%$prob_classes);
    } else {
        return $classes->[$pos-1];
    }
}

sub _load_file {    ## XXXX - later might be useful to accept encoding
    my $file = shift;
    my $str = "";
    open my $fh, "<:utf8", $file or die "Can not open file $file for reading: $!";
    while (<$fh>) {
        $str .= $_;
    }
    close $fh;

    return $str;
}

sub _load_thetas {
    my $path = shift;

    my $dir;

    opendir $dir, $path;
    my @ts = readdir $dir;

    for my $tfile (@ts) {
        if ($tfile =~ /theta-(\d+)\.dat/) {
            my $file = catfile($path, $tfile);
            print STDERR "Loading '$file'\n";
            $thetas->[$1 - 1] = Matrix->read($file);
        }
    }

    closedir $dir;
}

sub _compute_features {
    my $str = shift;

    my $alphabets = Lingua::Identifier::Feature::Alphabet::features($str);
    my $trigrams  = Lingua::Identifier::Feature::Trigrams::features($str);

    return { %$trigrams, %$alphabets };
}

=head1 AUTHOR

Alberto Simões, C<< <ambs at cpan.org> >>

=head1 ACKNOWLEDGMENTS

=over 4

=item * Simon D. Byers

=back

=head1 BUGS

Please report any bugs or feature requests to C<bug-lingua-Identifier
at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Lingua-Identifier>.
I will be notified, and then you'll automatically be notified of
progress on your bug as I make changes.

=head1 SUPPORT

You look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Lingua-Identifier>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Lingua-Identifier>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Lingua-Identifier>

=item * Search CPAN

L<http://search.cpan.org/dist/Lingua-Identifier/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Alberto Simões.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Lingua::Identifier


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.020.
use strict;
use warnings;

use Module::Build 0.3601;
use lib qw{inc}; use MyBuilder;

my %module_build_args = (
  "build_requires" => {
    "Math::Matrix::MaybeGSL" => "0.006",
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Math::Matrix::MaybeGSL" => "0.006",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "A NN based approach for language identification",
  "dist_author" => [
    "Alberto Simoes <ambs\@cpan.org>"
  ],
  "dist_name" => "Lingua-Identifier",
  "dist_version" => "0.01_8",
  "license" => "perl",
  "module_name" => "Lingua::Identifier",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "File::ShareDir" => 0,
    "File::Spec::Functions" => 0,
    "Math::GSL" => 0,
    "Math::Matrix::MaybeGSL" => "0.006",
    "Text::Ngram" => 0,
    "perl" => "5.014004",
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "script_files" => [
    "bin/identify_language"
  ],
  "share_dir" => {
    "dist" => "share"
  },
  "test_requires" => {
    "Test::More" => 0
  }
);


my %fallback_build_requires = (
  "Math::Matrix::MaybeGSL" => "0.006",
  "Module::Build" => "0.3601",
  "Test::More" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = MyBuilder->new(%module_build_args);

$build->create_build_script;
